File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
'''
https://www.acmicpc.net/problem/2805
binary_search 나무자르기
'''
import sys
input = sys.stdin.readline
N, M = map(int, input().rstrip().split())
lengths = list(map(int, input().rstrip().split()))

start = 1
end = max(lengths)
answer = 0
while start <= end :
    mid = (start + end) // 2

    total = 0
    for len in lengths :
        if len > mid :
            total += len - mid
    
    # 필요한 길이 보다 크거나 같다. 절단기의 높이를 올려준다.
    if M <= total :
        answer = mid
        start = mid +1
    else :
        end = mid -1




print(answer)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.cs
using System;
using System.Linq;
namespace CodingTest
{
    internal class Program
    {
        static void Main(string[] args)
        {
            int[] input = Console.ReadLine().Split(" ").Select(int.Parse).ToArray();
            int M = input[1];
            int[] heights = Console.ReadLine().Split(" ").Select(int.Parse).ToArray();
            int start = 0;
            int end = heights.Max();
            int answer = 0;
            while(start <= end)
            {
                long sum = 0;
                int mid = (start + end) / 2;
                for(int i=0; i<heights.Length; i++)
                {
                    sum += Math.Max(0, heights[i] - mid);
                }
                if(sum >= M)
                {
                    answer = mid;
                    start = mid + 1;
                }
                else
                {
                    end = mid - 1;
                }
            }
            Console.WriteLine(answer);
        }
    }
}
